src/category_theory/isomorphism.lean

```diff
@@ -157,41 +159,63 @@ by { erw [inv_eq_inv α.symm β, eq_comm], refl }
 
 end iso
 
-/-- `is_iso` typeclass expressing that a morphism is invertible.
-    This contains the data of the inverse, but is a subsingleton type. -/
-class is_iso (f : X ⟶ Y) :=
-(inv : Y ⟶ X)
-(hom_inv_id' : f ≫ inv = 𝟙 X . obviously)
-(inv_hom_id' : inv ≫ f = 𝟙 Y . obviously)
+/-- `is_iso` typeclass expressing that a morphism is invertible. -/
+def is_iso (f : X ⟶ Y) : Prop := ∃ inv : Y ⟶ X, f ≫ inv = 𝟙 X ∧ inv ≫ f = 𝟙 Y
+
+attribute [class] is_iso
+
+/--
+The inverse of a morphism `f` when we have `[is_iso f]`.
+-/
+noncomputable def inv (f : X ⟶ Y) [I : is_iso f] := classical.some I
 
-export is_iso (inv)
+namespace is_iso
+
+@[simp, reassoc] lemma hom_inv_id (f : X ⟶ Y) [I : is_iso f] : f ≫ inv f = 𝟙 X :=
+(classical.some_spec I).left
+@[simp, reassoc] lemma inv_hom_id (f : X ⟶ Y) [I : is_iso f] : inv f ≫ f = 𝟙 Y :=
+(classical.some_spec I).right
+
+end is_iso
+
+open is_iso
 
 /-- Reinterpret a morphism `f` with an `is_iso f` instance as an `iso`. -/
-def as_iso (f : X ⟶ Y) [h : is_iso f] : X ≅ Y := { hom := f, ..h }
+noncomputable
+def as_iso (f : X ⟶ Y) [h : is_iso f] : X ≅ Y := ⟨f, inv f, hom_inv_id f, inv_hom_id f⟩
 
 @[simp] lemma as_iso_hom (f : X ⟶ Y) [is_iso f] : (as_iso f).hom = f := rfl
 @[simp] lemma as_iso_inv (f : X ⟶ Y) [is_iso f] : (as_iso f).inv = inv f := rfl
 
 namespace is_iso
 
-@[simp] lemma hom_inv_id (f : X ⟶ Y) [is_iso f] : f ≫ inv f = 𝟙 X :=
-is_iso.hom_inv_id'
-@[simp] lemma inv_hom_id (f : X ⟶ Y) [is_iso f] : inv f ≫ f = 𝟙 Y :=
-is_iso.inv_hom_id'
+@[priority 100] -- see Note [lower instance priority]
+instance epi_of_iso (f : X ⟶ Y) [is_iso f] : epi f  :=
+{ left_cancellation := λ Z g h w,
+  -- This is an interesting test case for better rewrite automation.
+  by rw [← is_iso.inv_hom_id_assoc f g, w, is_iso.inv_hom_id_assoc f h] }
+@[priority 100] -- see Note [lower instance priority]
+instance mono_of_iso (f : X ⟶ Y) [is_iso f] : mono f :=
+{ right_cancellation := λ Z g h w,
+  by rw [← category.comp_id g, ← category.comp_id h, ← is_iso.hom_inv_id f, ← category.assoc, w,
+    ← category.assoc] }
 
-@[simp] lemma hom_inv_id_assoc {Z} (f : X ⟶ Y) [is_iso f] (g : X ⟶ Z) :
-  f ≫ inv f ≫ g = g :=
-(as_iso f).hom_inv_id_assoc g
+@[ext] lemma ext {f : X ⟶ Y} [is_iso f] {g : Y ⟶ X}
+  (hom_inv_id : f ≫ g = 𝟙 X) : inv f = g :=
+begin
+  apply (cancel_epi f).mp,
+  simp [hom_inv_id],
+end
 
-@[simp] lemma inv_hom_id_assoc {Z} (f : X ⟶ Y) [is_iso f] (g : Y ⟶ Z) :
-  inv f ≫ f ≫ g = g :=
-(as_iso f).inv_hom_id_assoc g
+@[ext] lemma ext' {f : X ⟶ Y} [is_iso f] {g : Y ⟶ X}
+  (hom_inv_id : f ≫ g = 𝟙 X) : g = inv f :=
```

Comment:
Perhaps it would be nice to have a lemma like this with the assumption `g ≫ f = 𝟙 Y` instead? Converting between your `ext'` and `ext` isn't too hard but it seems less simple to get `inv f = g` from `g ≫ f = 𝟙 Y` using the current API.